src/parser/jobs/sam/modules/Shoha.tsx

```diff
@@ -0,0 +1,272 @@
+		const generatedStacks = GENERATORS[abilityId]
+
+		this.addGeneratedStackAndPush(generatedStacks, abilityId)
+	}
+	}
+
+		// This is exclusively for stacks gained via Meditate channel. and not any stacks gained from GCD casts
+	private onMeditateGain() {
+			this.stacks++
+			this.totalGeneratedStacks++
+
+			// just a safety net
+			if (this.stacks > MAX_STACKS) {
+				this.stacks = MAX_STACKS
+			}
+
+			this.pushToHistory()
+	}
+
+	private addGeneratedStackAndPush(generatedStacks: number, abilityId: number) {
+		this.stacks += generatedStacks
+		this.totalGeneratedStacks += generatedStacks
+		if (this.stacks > MAX_STACKS) {
+			const waste = this.stacks - MAX_STACKS
+			this.wasteBySource[abilityId] += waste
+			this.stacks = MAX_STACKS
+		}
+
+		this.pushToHistory()
+	}
+
+	private onSpender(event: CastEvent) {
+		this.stacks = this.stacks - SPENDERS[event.ability.guid]
+		this.shohaUses++
+
+		// safety net!
+
+		if (this.stacks < 0) {
+			this.stacks = 0
+		}
+
+		this.pushToHistory()
+
+	}
+
+	private onDeath() {
+		this.wasteBySource[ACTIONS.RAISE.id] += this.stacks
+		this.dumpRemainingResources()
+	}
+
+	private dumpRemainingResources() {
+		this.stacks = 0
+		this.pushToHistory()
+	}
+
+	private pushToHistory() {
+		const timestamp = this.parser.currentTimestamp - this.parser.fight.start_time
+		this.stackHistory.push({t: timestamp, y: this.stacks})
+	}
+
+	private onComplete() {
+		this.dumpRemainingResources()
+
+		// final use amounts
+
+		const totalPossibleUses = Math.floor(this.totalGeneratedStacks/MAX_STACKS)
+		const totalUses = this.shohaUses
+
+		this.checklist.add(new Rule({
+			name: 'Meditation',
+			description: <Trans id="sam.shoha.waste.content">
+				Wasted meditation generation, ending the fight with stacks fully charged, or dying with stacks charged is a
+				direct potency loss. Use <ActionLink {...ACTIONS.SHOHA}/> to avoid wasting stacks.
+			</Trans>,
+			requirements: [
+				new Requirement({
+					name: <Trans id="sam.shoha.checklist.requirement.waste.name">
+						Use as many of your meditation stacks as possible.
+					</Trans>,
+					value: totalUses,
+					target: totalPossibleUses,
+				}),
+			],
+		}))
+	}
+
+	private convertWasteMapToTable() {
+		const rows = [
+			this.convertWasteEntryToRow(ACTIONS.HIGANBANA),
+			this.convertWasteEntryToRow(ACTIONS.TENKA_GOKEN),
+			this.convertWasteEntryToRow(ACTIONS.MIDARE_SETSUGEKKA),
+			this.convertWasteEntryToRow(ACTIONS.KAESHI_HIGANBANA),
+			this.convertWasteEntryToRow(ACTIONS.KAESHI_GOKEN),
+			this.convertWasteEntryToRow(ACTIONS.KAESHI_SETSUGEKKA),
+			this.convertWasteEntryToRow(ACTIONS.RAISE),
+		]
+
+		return <Fragment key="wasteBySource-fragment">
+			<table key="wasteBySource-table">
+				<tbody key="wasteBySource-tbody">
+					{rows}
+				</tbody>
+			</table>
+		</Fragment>
+	}
+
+	private convertWasteEntryToRow(action: TODO) {
+		let actionName = action.name
+		if (action === ACTIONS.RAISE) {
+			actionName = 'Death'
```

Comment:
This will probably break i18n but I never found a good workaround for that when I did similar in MNK either tbh.

Reply:
Pretty sure this isn't actually going to be doing anything.

`<ActionLink name={actionName} {...action}/>` can essentially be thought of as `{name: actionName, ...action}` - given `action` already _has_ a name, action's spread will override the earlier assignment.

Can technically use `<ActionLink {...action} name={actionName}/>` to do it the other way around, but that won't work either - the `name` is only used while it loads data off the API, at which point API data takes precedence.

If you _really_ want to show "death" instead of the action name for raise, you can do something like uh (and this is ugly)

```tsx
const actionChildren = action === ACTIONS.RAISE
	? <Trans id="some.id.here">Death</Trans>
	: undefined

return <>...<ActionLink {...action} children={actionChildren}/>...</>
```